### ISSUE TITLE
Make the command whitelisting feature optional

### ISSUE THREAD

**Issue by ricardolsmendes** [square/rce-agent#33]
Hi there,
We have been discussing the possibility of using __rce-agent__ to add a _container-to-container communication feature_ to [GitLab Runner's Custom Executor driver for AWS Fargate](https://gitlab.com/gitlab-org/ci-cd/custom-executor-drivers/fargate). The related issue is [Support a fargate-agent sidecar container to prevent the need for custom CI images](https://gitlab.com/gitlab-org/ci-cd/custom-executor-drivers/fargate/-/issues/12).

We made good progress in the proof-of-concept, but now we are blocked by the __rce-agent__'s command whitelisting feature. This is a blocker because the driver is used to run CI/CD pipelines and we cannot anticipate the commands users will set up in their jobs, and hence we cannot whitelist them.

So the primary goal of this issue is to ask __rce-agent__'s team if it makes sense to turn the command whitelisting feature optional. Based on the answer we can decide whether it's worth keeping the conversation going. 🙂 

Thanks in advance!

**Comment by daniel-nichter** [comment]
Hi @ricardolsmendes. We discussed and yes, we can do this. As you mentioned, we can make it an option that users have to explicitly enable (so the code defaults to most secure: deny by default). One question: do you need just to allow all commands, or do you want that plus also a block list? (I'm not sure if a block list would be much use because there are countless commands one could block. Maybe it'd be useful just to block certain known risky commands like "shutdown".)

**Comment by ricardolsmendes** [comment]
Thanks, @daniel-nichter! That's really great news! 🙂

I agree command whitelisting should be enabled by default as it increases security. Answering your question, the block list wouldn't be useful for us in the current use case. GitLab Runner _enriches_ users' commands (by prepending environment variables, for example), so the server will never receive a simple command such as `shutdown`.

Regarding the development workflow, I understood Square's team will take care of this improvement:

>  yes, we can do this

Am I right?

**Comment by daniel-nichter** [comment]
You're right. I'll try to implement this this week, time permitting.

**Comment by ricardolsmendes** [comment]
Thanks in advance, @daniel-nichter!

**Comment by daniel-nichter** [comment]
@ricardolsmendes Want to give https://github.com/square/rce-agent/pull/34 a try? You'll need to use new `ServerConfig` and `NewServerWithConfig`. I'm writing more unit tests and waiting for another Square engineer to review.

**Comment by ricardolsmendes** [comment]
Thanks, @daniel-nichter. I'm planning to give it a try later today!

**Comment by ricardolsmendes** [comment]
Hi @daniel-nichter, I'm facing issues with the new code. Let me describe what I've done so far:

1. Created an example server that allows any command: https://github.com/ricardolsmendes/rce-agent/blob/server-any-command/example/server-any-command/main.go.
1. Started it by the command `./server-any-command --allow-any-command --disable-security`.
1. Using the existing example client, I ran `./client go version` and got an error:
```plaintext
2020/07/14 18:47:51 Connecting to 127.0.0.1:5501...
2020/07/14 18:47:51 Connected
       ID: 58145f4ed1184c1288cb820a9357cb7d
     Name: go
    State: FAIL
      PID: 19785
StartTime: 1594763271131489188
 StopTime: 1594763271143900477
 ExitCode: 2
    Error: 
   Stdout: 
   Stderr: 
         : go go: unknown command
         : Run 'go help' for usage.
```
1. The server log shows:
```plaintext
2020/07/14 21:47:51.131213 server.go:201: cmd=58145f4ed1184c1288cb820a9357cb7d: start: go path: go args: [go version]
2020/07/14 21:47:51.134263 server.go:208: cmd=58145f4ed1184c1288cb820a9357cb7d: wait
2020/07/14 21:47:51.143975 repo.go:52: cmd=58145f4ed1184c1288cb820a9357cb7d remove
2020/07/14 21:47:51.144029 server.go:227: cmd=58145f4ed1184c1288cb820a9357cb7d: wait return
```
I am running the client and the server locally and have Go installed here: `go version go1.14.2 linux/amd64`.

Thoughts? :thinking:

**Comment by daniel-nichter** [comment]
Pushed a fix. The unit test for this wasn't so good, so updated and fixed.

Strangely, though, I see the commit everywhere but the PR: https://github.com/square/rce-agent/pull/34. There should be 4 commits on that PR, but GitHub isn't updating the PR. GitHub shows the 4th commit on the branch, though.  If you pull the branch you should get commit 13704e2763e785834e166597dc86f96211f6d945 ("Fix and better test for any command").

**Comment by ricardolsmendes** [comment]
Great, thanks! My local tests work as expected with the fix. Now I'll try more complex scenarios and let you know if I face any issue. Best,

**Comment by daniel-nichter** [comment]
PR #34 merged (v1.1.0). If any issues with this new stuff, please open a new issue. Thanks!
